geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/key/AbstractDumpRestoreIntegrationTest.java

```diff
@@ -0,0 +1,241 @@
+  private JedisCluster jedis;
+  private static String STRING_VALUE;
+  private static byte[] RESTORE_BYTES;
+
+  @BeforeClass
+  public static void setupClass() {
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SET_ID,
+        RedisSet.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_STRING_ID,
+        RedisString.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_HASH_ID,
+        RedisHash.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SORTED_SET_ID,
+        RedisSortedSet.class);
+  }
+
+  @Before
+  public void setup() {
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, getPort()), REDIS_CLIENT_TIMEOUT);
+
+    RedisClusterClient client =
+        RedisClusterClient.create(String.format("redis://%s:%d", BIND_ADDRESS, getPort()));
+    lettuce = client.connect().sync();
+
+    STRING_VALUE = "It's a mad, mad, mad, mad, mad world";
+    lettuce.set("set-dump-value", STRING_VALUE);
+    RESTORE_BYTES = lettuce.dump("set-dump-value");
+  }
+
+  @After
+  public void tearDown() {
+    flushAll();
+  }
+
+  @Test
+  public void dumpTakesExactlyOneArgument() {
+    assertExactNumberOfArgs(jedis, Protocol.Command.DUMP, 1);
+  }
+
+  @Test
+  public void restoreErrorsWithUnknownOption() {
+    assertThatThrownBy(
+        () -> jedis.sendCommand("key", Protocol.Command.RESTORE, "key", "0", "", "FOO"))
+            .hasMessageContaining("ERR syntax error");
+  }
+
+  @Test
+  public void restoreFails_whenKeyAlreadyExists() {
+    lettuce.set("restored", "already exists");
+
+    assertThatThrownBy(() -> lettuce.restore("restored", 0, RESTORE_BYTES))
+        .hasMessage("BUSYKEY Target key name already exists.");
+  }
+
+  @Test
+  public void restoreFails_whenTTLisNegative() {
+    assertThatThrownBy(() -> lettuce.restore("restored", -1, RESTORE_BYTES))
+        .hasMessage("ERR Invalid TTL value, must be >= 0");
+  }
+
+  @Test
+  public void restoreFails_withInvalidBytes() {
+    assertThatThrownBy(() -> lettuce.restore("restored", 0, new byte[] {0, 1, 2, 3}))
+        .hasMessage("ERR DUMP payload version or checksum are wrong");
+  }
+
+  @Test
+  public void dumpAndRestoreString() {
+    lettuce.set("dumped", STRING_VALUE);
+
+    byte[] rawBytes = lettuce.dump("dumped");
+    lettuce.restore("restored", 0, rawBytes);
+    String response = lettuce.get("restored");
+
+    assertThat(response).isEqualTo(STRING_VALUE);
+    assertThat(lettuce.ttl("restored")).isEqualTo(-1);
+  }
+
+  @Test
+  public void restore_withTTL_setsTTL() {
+    lettuce.restore("restored", 2000, RESTORE_BYTES);
+    String response = lettuce.get("restored");
+
+    assertThat(response).isEqualTo(STRING_VALUE);
+    assertThat(lettuce.pttl("restored"))
+        .isLessThan(2000)
```

Comment:
It seems unlikely, but could this test possibly fail due to reaching this assertion before the ttl has had time to decrease from the initially set value? This assertion could then be `.isLessThatOrEqualTo(2000)` to guard against that.

Reply:
Yes, it is unlikely but I've changed it regardless.